File: go/pkg/mechbase/assets.go
package mechbase

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
)

// AssetInput is the create/upsert/update body. Nil fields are omitted.
type AssetInput struct {
	Name              *string `json:"name,omitempty"`
	ExternalID        *string `json:"external_id,omitempty"`
	SectionID         *int    `json:"section_id,omitempty"`
	SectionExternalID *string `json:"section_external_id,omitempty"`
	ZoneID            *int    `json:"zone_id,omitempty"`
	ZoneExternalID    *string `json:"zone_external_id,omitempty"`
	EquipmentType     *string `json:"equipment_type,omitempty"`
	MachineClass      *string `json:"machine_class,omitempty"`
}

// Assets is the asset registry resource scoped to an installation.
type Assets struct {
	client *Client
	iid    int
}

// ListAssetsOptions filters Assets.List. Zero values are omitted.
type ListAssetsOptions struct {
	Q      string
	ZoneID int
	Limit  int
	Offset int
}

// List returns a page of assets.
func (a *Assets) List(ctx context.Context, opts ListAssetsOptions) ([]Asset, error) {
	q := url.Values{}
	if opts.Q != "" {
		q.Set("q", opts.Q)
	}
	if opts.ZoneID != 0 {
		q.Set("zone_id", fmt.Sprintf("%d", opts.ZoneID))
	}
	addPaging(q, opts.Limit, opts.Offset)
	var env listEnvelope[Asset]
	if err := a.client.doJSON(ctx, http.MethodGet, pathf(a.iid, "/assets"), q, nil, &env); err != nil {
		return nil, err
	}
	return env.Items, nil
}

// Get fetches one asset by numeric id.
func (a *Assets) Get(ctx context.Context, assetID int) (*Asset, error) {
	var out Asset
	if err := a.client.doJSON(ctx, http.MethodGet, pathf(a.iid, fmt.Sprintf("/assets/%d", assetID)), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Create creates an asset.
func (a *Assets) Create(ctx context.Context, in AssetInput) (*Asset, error) {
	var out Asset
	if err := a.client.doJSON(ctx, http.MethodPost, pathf(a.iid, "/assets"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Upsert creates or updates an asset, matched by external_id.
func (a *Assets) Upsert(ctx context.Context, in AssetInput) (*Asset, error) {
	var out Asset
	if err := a.client.doJSON(ctx, http.MethodPut, pathf(a.iid, "/assets"), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update partially updates an asset by numeric id.
func (a *Assets) Update(ctx context.Context, assetID int, in AssetInput) (*Asset, error) {
	var out Asset
	if err := a.client.doJSON(ctx, http.MethodPatch, pathf(a.iid, fmt.Sprintf("/assets/%d", assetID)), nil, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes an asset; cascade also removes children.
func (a *Assets) Delete(ctx context.Context, assetID int, cascade bool) (*DeleteResult, error) {
	q := url.Values{}
	if cascade {
		q.Set("cascade", "true")
	}
	var out DeleteResult
	if err := a.client.doJSON(ctx, http.MethodDelete, pathf(a.iid, fmt.Sprintf("/assets/%d", assetID)), q, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: swift/Sources/Mechbase/Assets.swift
import Foundation

/// Create/upsert/update body for assets. Nil fields are omitted by the encoder
/// (the SDK JSONEncoder skips nil optionals).
public struct AssetInput: Encodable, Sendable {
    public var name: String?
    public var externalId: String?
    public var sectionId: Int?
    public var sectionExternalId: String?
    public var zoneId: Int?
    public var zoneExternalId: String?
    public var equipmentType: String?
    public var machineClass: String?

    public init(name: String? = nil, externalId: String? = nil, sectionId: Int? = nil,
                sectionExternalId: String? = nil, zoneId: Int? = nil, zoneExternalId: String? = nil,
                equipmentType: String? = nil, machineClass: String? = nil) {
        self.name = name; self.externalId = externalId; self.sectionId = sectionId
        self.sectionExternalId = sectionExternalId; self.zoneId = zoneId
        self.zoneExternalId = zoneExternalId; self.equipmentType = equipmentType
        self.machineClass = machineClass
    }

    enum CodingKeys: String, CodingKey {
        case name
        case externalId = "external_id"
        case sectionId = "section_id"
        case sectionExternalId = "section_external_id"
        case zoneId = "zone_id"
        case zoneExternalId = "zone_external_id"
        case equipmentType = "equipment_type"
        case machineClass = "machine_class"
    }
}

public struct Assets: Sendable {
    let scope: InstallationScope
    init(http: HTTPClient, installationId: Int) {
        self.scope = InstallationScope(http: http, installationId: installationId)
    }

    public func list(query: String? = nil, zoneId: Int? = nil, limit: Int = 50, offset: Int = 0) async throws -> [Asset] {
        let q: [URLQueryItem] = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "zone_id", value: zoneId.map(String.init)),
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "offset", value: String(offset)),
        ]
        return try await scope.http.get(scope.path("/assets"), query: q, as: Page<Asset>.self).items
    }

    public func get(id: Int) async throws -> Asset {
        try await scope.http.get(scope.path("/assets/\(id)"), as: Asset.self)
    }

    public func create(_ input: AssetInput) async throws -> Asset {
        try await scope.http.postJSON(scope.path("/assets"), body: input, as: Asset.self)
    }

    public func upsert(_ input: AssetInput) async throws -> Asset {
        try await scope.http.putJSON(scope.path("/assets"), body: input, as: Asset.self)
    }

    public func update(id: Int, _ input: AssetInput) async throws -> Asset {
        try await scope.http.patchJSON(scope.path("/assets/\(id)"), body: input, as: Asset.self)
    }

    public func delete(id: Int, cascade: Bool = false) async throws -> DeleteResult {
        let q = cascade ? [URLQueryItem(name: "cascade", value: "true")] : []
        return try await scope.http.delete(scope.path("/assets/\(id)"), query: q, as: DeleteResult.self)
    }
}
